pkg/vsphere/disk/disk_manager.go

```diff
@@ -303,23 +325,28 @@ func (m *Manager) Detach(op trace.Operation, d *VirtualDisk) error {
 	})
 	m.reconfig.Unlock()
 
+	return err
+}
+
+// detachAll detaches all disks from this vm
+func (m *Manager) detachAll(op trace.Operation) error {
+	disks, err := findAllDisks(op, m.vm)
 	if err != nil {
-		op.Errorf("detach for %s failed with %s", d.DevicePath, errors.ErrorStack(err))
-		return errors.Trace(err)
+		return err
 	}
 
-	func() {
-		select {
-		case <-m.maxAttached:
-		default:
+	for _, disk := range disks {
+		if err = m.detach(op, disk); err != nil {
+			op.Errorf("error detaching disk: %s", err.Error())
```

Comment:
I'm not sure what the right path is;  exit early like the code does or exit later after trying all options.  

But, I think you may be right.  Detatching late is probably what the caller wants.  I'll fix this.  Thanks!